src/group_theory/submonoid.lean

```diff
@@ -657,260 +623,516 @@ instance : has_Inf (submonoid M) :=
+
+open submonoid
+
+/-- The range of a monoid homomorphism is a submonoid. -/
+@[to_additive "The range of an `add_monoid_hom` is an `add_submonoid`."]
+def range (f : M →* N) : submonoid N := (⊤ : submonoid M).map f
+
+@[simp, to_additive] lemma coe_range (f : M →* N) :
+  (f.range : set N) = set.range f :=
+set.image_univ
+
+@[simp, to_additive] lemma mem_range {f : M →* N} {y : N} :
+  y ∈ f.range ↔ ∃ x, f x = y :=
+by simp [range]
+
+@[to_additive]
+lemma map_range (g : N →* P) (f : M →* N) : f.range.map g = (g.comp f).range :=
+(⊤ : submonoid M).map_map g f
 
 /-- Restriction of a monoid hom to a submonoid of the domain. -/
 @[to_additive "Restriction of an add_monoid hom to an `add_submonoid` of the domain."]
-def restrict {N : Type*} [monoid N] (f : M →* N) : S →* N :=
-⟨λ s, f s, f.map_one, λ x y, f.map_mul x y⟩
+def restrict {N : Type*} [monoid N] (f : M →* N) : S →* N := f.comp S.subtype
+
+@[to_additive]
+lemma restrict_apply {N : Type*} [monoid N] (f : M →* N) (x : S) : f.restrict S x = f x := rfl
 
 /-- Restriction of a monoid hom to a submonoid of the codomain. -/
 @[to_additive "Restriction of an `add_monoid` hom to an `add_submonoid` of the codomain."]
-def subtype_mk {N : Type*} [monoid N] (f : N →* M) (h : ∀ x, f x ∈ S) : N →* S :=
+def cod_restrict {N : Type*} [monoid N] (f : N →* M) (h : ∀ x, f x ∈ S) : N →* S :=
 { to_fun := λ n, ⟨f n, h n⟩,
   map_one' := subtype.eq f.map_one,
   map_mul' := λ x y, subtype.eq (f.map_mul x y) }
 
 /-- Restriction of a monoid hom to its range. -/
 @[to_additive "Restriction of an `add_monoid` hom to its range."]
-def range_mk {N} [monoid N] (f : M →* N) : M →* f.range :=
-subtype_mk f.range f $ λ x, ⟨x, submonoid.mem_top x, rfl⟩
+def range_restrict {N} [monoid N] (f : M →* N) : M →* f.range :=
+f.cod_restrict f.range $ λ x, ⟨x, submonoid.mem_top x, rfl⟩
+
+@[to_additive]
+lemma range_top_iff_surjective {N} [monoid N] {f : M →* N} :
+  f.range = (⊤ : submonoid N) ↔ function.surjective f :=
+submonoid.ext'_iff.trans $ iff.trans (by rw [coe_range, coe_top]) set.range_iff_surjective
 
 /-- The range of a surjective monoid hom is the whole of the codomain. -/
 @[to_additive "The range of a surjective `add_monoid` hom is the whole of the codomain."]
 lemma range_top_of_surjective {N} [monoid N] (f : M →* N) (hf : function.surjective f) :
   f.range = (⊤ : submonoid N) :=
-submonoid.ext'_iff.1 $ (set.ext_iff _ _).2 $ λ x,
-⟨λ h, submonoid.mem_top x, λ h, exists.elim (hf x) $ λ w hw, ⟨w, submonoid.mem_top w, hw⟩⟩
+range_top_iff_surjective.2 hf
 
-/-- The monoid hom associated to an inclusion of submonoids. -/
-@[to_additive "The `add_monoid` hom associated to an inclusion of submonoids."]
-def set_inclusion (T : submonoid M) (h : S ≤ T) : S →* T :=
-subtype_mk _ S.subtype (λ x, h x.2)
+/-- Multiplicative kernel of a monoid hom as a submonoid of its domain-/
+@[to_additive "Additive kernel of an `add_monoid_hom` as a submonoid of its domain"]
+def ker (f : M →* N) := (⊥ : submonoid N).comap f
```

Comment:
Is this definition useful? The set of things a monoid hom sends to 1 doesn't determine its image. The more useful notion, ```eq_locus```, gets called 'Ker' in universal algebra stuff so I followed that in ```group_theory.congruence```; maybe I should rename ```ker``` in that file to ```eq_locus``` to avoid confusion, and we keep ```monoid_hom.ker``` for group homs?

Reply:
Good question! I've never worked with this `eq_locus` before. But maybe that's what we need.

Reply:
Let's keep the name `ker` here, perhaps adding a comment to the effect that it's a bad notation for monoids. Let's rename the good notion for monoids to `eq_locus`, as @101damnations suggests.

Reply:
I think that having similar API for monoids, groups, modules etc. is important. if this definition is not useful for monoids, we can just delete it.

Reply:
Ok, on second thoughts let's delete this then.

Reply:
I would also vote for deletion. We are going to run into this when we go back to https://github.com/leanprover-community/mathlib/pull/2140 , which has a conflicting definition of `ker`. The alternative fix is to call the group kernel `gker` which is kind of horrible as it will be the kernel which is by far the most popular. In 2140 there is also `comap`, `map` and `range`. Here there will have to be some decision abut whether we go for e.g. `gcomap` for groups.

Reply:
`map` and `comap` should go into `submonoid` and `subgroup` namespaces. We'll delete `ker`, so the only potential conflicts will be `range` and `eq_locus`. Should I use `m` prefix for monoids?

Reply:
I think the `m` prefix is fine.